### ISSUE TITLE
Building doesn't render on top of highway areas any more

### ISSUE THREAD

**Issue by sb12** [gravitystorm/openstreetmap-carto#688]
Example way: http://www.openstreetmap.org/way/104978884
Other example area: http://www.openstreetmap.org/#map=19/52.52496/13.36941

This happened after the recent style changes a few days ago. 
Before buildings and roofs were rendered on top of the highway areas (and highway ways with covered=yes/tunnel=yes) and you could still see the highways because of the building transparency. Now you can't see the buildings anymore when there's a highway area inside.
Is this a wanted behaviour or an error?

**Comment by kocio-pl** [comment]
So - are there any ideas how to change the stylesheet? I think it's getting serious after such a long time because such places (with complicated buildings/highways layers) are probably important.

**Comment by daganzdaanda** [comment]
IMHO we could start by rendering anything that is tagged covered=yes with 50% opacity. No need to change rendering order then. It might still look weird, but it should make it more clear what's going on. 

Later, other changes may be smarter, like interpreting 3D-tags or the reative layers, or "location".

**Comment by kocio-pl** [comment]
Could someone make a test renderings of it? I suggest Berlin Hauptbahnhof as a great example of complex multilevel building, but I'm also very interested in Dworzec Centralny in Warsaw (my hometown).

**Comment by jojo4u** [comment]
Is this building (http://www.openstreetmap.org/way/23691951) being under rather over a railway=platform layer/level=-3 the same problem or a different issue?

**Comment by Theodin** [comment]
It is over the platform asn the building is on the surface and the platform is a subway platform with 2 moving stairways leading down.

**Comment by kocio-pl** [comment]
It looks like the hack with render=no [stopped working](http://www.openstreetmap.org/way/308077768) and Eiffel Tower is now broken too. 

@gravitystorm Andy, so what about such important places? Do we still accept this class of problems as unavoidable cost of better showing the pedestrian areas (as we do it now) or maybe we should have kind of workaround for such exceptions?

**Comment by matkoniecz** [comment]
"render=no" never had any effect on whatever something will be rendered.

**Comment by kocio-pl** [comment]
So maybe that was something else - I just know that building was visible back then, but the problem is what in general should we do with such things?

**Comment by malenki** [comment]
Every time I see an amenity=fuel with a building=roof and layer=1 with service highways going over them I sigh inwardly.
Now I even bothered to search for this issue to write this comment.
A solution would be nice.

**Comment by HolgerJeromin** [comment]
@malenki tag the service roads with covered=yes and they will look nicer.

**Comment by jojo4u** [comment]
> Edit: i just added this "hint" to the English and German fuel wikipage 

No, not yet. This is undiscussed since layer=* is technically speaking enough. This is a shortcoming of the renderer. See #1710

**Comment by HolgerJeromin** [comment]
covered is Status: Approved 
And  it is not wrong from the data point of view. If it helps current rendering it is not wrong!

**Comment by jojo4u** [comment]
I opened a topic on the talk page: https://wiki.openstreetmap.org/wiki/Talk:Key:covered#What_to_do_with_building.3Droof.3F

**Comment by mboeringa** [comment]
> This is undiscussed since layer=* is **technically** *speaking enough*.

I think it is more appropriate to say that it is ***theoretically*** **speaking enough**.

The **technical** implications of having to properly render buildings or other polygons over or under roads, *or any other line feature*, based on the **layer=x** tag, is far more complicated, and AFAIK, hasn't been really implemented in any renderer yet, except maybe those fully focusing on 3D, like F4Map...

**Comment by sladen** [comment]
I don't think this can be batted away on the basis of requiring **layer=x**, as the Effiel Tower (with **layer=4**) demonstrates:

* http://www.openstreetmap.org/way/5013364

**Comment by kocio-pl** [comment]
I find the Eiffel Tower as a major bug of new policy (highways always on top of buildings) cryptographically wise.

**Comment by daganzdaanda** [comment]
In case of the Eiffel Tower, the highway=pedestrian with covered=yes http://www.openstreetmap.org/way/308077768 should be rendered differently. Since transparency is evil, we could introduce another solid colour (darker?). Following this with all the other things that could be covered, we would need a lot of new colours!! Seems unlikely to be possible without conflicts with existing colours.

So what about @mboeringa s idea of a hatching / chessboard pattern of 100% solid colour? (He does it with roofs, can't find a link) 
In the "empty" areas between the solid colour, one will see the thing that is physically on top of the covered area. This way, one could see the pedestrian area and the Eiffel Tower at the same time. 
It may be a bit noisy visually, depending on the size of the pattern.

**Comment by kocio-pl** [comment]
Interesting idea to test. I would like to see this example link very much.

**Comment by daganzdaanda** [comment]
Now I found it: https://github.com/gravitystorm/openstreetmap-carto/issues/327#issuecomment-85951265
![roof example](https://cloud.githubusercontent.com/assets/7635726/5050143/0991c4ea-6c2c-11e4-9503-499104e354c4.jpg)

**Comment by kocio-pl** [comment]
I guess this is done in ArcGIS. Is it possible to do with osm-carto tools?

**Comment by mboeringa** [comment]
Yes, this was ArcGIS, but why wouldn't a simple fine cross-hatch be possible? There are already single line hatches in use for danger areas and so...

The feature is as @daganzdaanda pointed out based on the **building=roof** tag, often employed at railway stations.

**Comment by kocio-pl** [comment]
I was thinking rather about chessboard, because I haven't seen something like this on osm-carto yet. Hatching is obviously possible.

And I don't get the details how it could be done with Eiffel Tower?

**Comment by daganzdaanda** [comment]
> And I don't get the details how it could be done with Eiffel Tower?

The pedestrian area under the tower is tagged correctly with "covered=yes". Because it's a "highway", we render it on top of the building. So my idea is to render the area with a crosshatch, so that the building will show through. That way, all highways can still be rendered on top of buildings, because we can make the parts that are "covered=yes" fake-transparent.

**Comment by dieterdreist** [comment]
2016-02-18 19:00 GMT+01:00 mboeringa <notifications@github.com>:

> Yes, this was ArcGIS, but why wouldn't a simple fine cross-hatch be
> possible?



AFAIK you can't currently do vector based hatches in mapnik, but have to
use the
PolygonPatternSymbolizerto do it.

**Comment by daganzdaanda** [comment]
If it works, a crosshatch might also be useful for showing that platforms are "covered" or underground though this will still be more complicated I guess...

**Comment by mboeringa** [comment]
> AFAIK you can't currently do vector based hatches in mapnik, but have to use the PolygonPatternSymbolizerto do it.

I don't know if "hatch" is the right word to use in the context of mapnik and carto, but in ArcGIS, this is indeed a vector based 'polygon' symbol type, actually a "Line Fill Symbol" to be precise, and two of these stacked on top of each other to create the final cross-hatch polygon fill symbol. So you can't define a cross-hatch directly in ArcGIS, but use a multi-layer symbol instead.

**Comment by matkoniecz** [comment]
There are following situations
 - There is a building above highway area ([Eiffel Tower](http://www.openstreetmap.org/#map=17/48.85740/2.29346) is one example, there is also https://github.com/gravitystorm/openstreetmap-carto/issues/688#issuecomment-47841979)
- there is an underground building below highway area (covered also by #552)
- there is a hole in highway area ([buildings on a square](http://www.openstreetmap.org/#map=18/50.06172/19.93858))

Rendering buildings above highway areas would fix the first issue (at cost of hiding highway area), give unexpected results for the second case and mask data error in a third case. It may also cause buildings to cover linear highway=* objects.

**Comment by sladen** [comment]
@daganzdaanda, what I don't quite follow here is that 50%/50% cross-hatching stippling exists because the world wasn't able to do proper transparency and blending.  And now that we have systems, tools, and file-formats that support proper transparency people are going out of their way to avoid using it.

For solving this we have metadata in the form of the **layer=** tag, and we have tools in the form of transparency and layering.  What I'm unclear about is why everyone appears to be going out of their way to avoid using the two together.

**Comment by matkoniecz** [comment]
> we have tools in the form of transparency and layering. What I'm unclear about is why everyone appears to be going out of their way to avoid using the two together.

Because layering multiple transparent objects gives unclear, unpredictable and ugly results in at least some cases. See for example https://github.com/gravitystorm/openstreetmap-carto/pull/590#issuecomment-44815238 https://github.com/gravitystorm/openstreetmap-carto/pull/590#issuecomment-44787176

Transparency has some uses like text halo but it works poorly for polygon fill.

**Comment by gorn** [comment]
I am not against the change in no layers are defined, but there is no reason to render it the same way is the layer of one object has higher positive value than the other one. That is the core meaning of the layer parameter anyway. I can not imagine any argument against this. What would be the meaning of layer otherwise?

**Comment by gorn** [comment]
In case on highway=pedestrina areas with layer=-1 this leads to really strange results. Basically all metro stations wil be rendered on top of all buildings. I do not think it is a wise decision. Se an example here #2113

**Comment by gorn** [comment]
The Eifell tower is still invisible. Althought I appreciate all technical difficulties with layering, you clearly underestimate negative marketing impact of letting it so for so long time. What impression  can anyone from "wide public" get if s/he looks at Eifel tower, any undeground station, bigger petrol station? It is very simple ... it is "crappy" for them and it will take another year or so until they will be willing to consider giving it second chance. 

So please revert it to the original state and only develop in in some technical branch until all difficult technical questions are discussed/solved. Also as Carto is used in most of the "wide public" applications it hurts the whole OpenStreetMap project.

**Comment by math1985** [comment]
>So please revert it to the original state 

Could somebody find out which commit caused this exactly?

**Comment by math1985** [comment]
> The Eifell tower is still invisible.

See https://github.com/gravitystorm/openstreetmap-carto/issues/172#issue-19634740 for the rationale behind the current rendering. In the older rendering, we violated requirement 2. In the new rendering, we violate requirement 1. As you will see, this is not an easy to solve problem.

**Comment by sladen** [comment]
math1985: Being functional probably wins over worries over tiny gaps in road casing.

_And arguably the 'gaps' in the road casing (drawn at actual size) actually improve the rendering by helping to balance the (generally) exaggerated width of the roads._

**Comment by kocio-pl** [comment]
@math1985: 

> Violating 1. is an option, but is undesirable as it would require large-scale re-tagging. All the highway-areas with a building in the middle to be tagged should be tagged as multipolygons if we decide to violate this requirement.

So it may be tedious job, but can be fixed, it's just "_undesirable_", while currently some things (like showing the outline of the Eiffel Tower) are simply _impossible_, which is worse for me.

**Comment by math1985** [comment]
> So it may be tedious job, but can be fixed,

We currently do have the option where we violate 1.

> There is also another possibility (at least theoretically): smaller areas could be visible over bigger areas.

I don't think that would make a difference here. Rendering any buildings on top of highway areas would imply, assuming highway areas are rendered on top of roads, that those buildings are obscuring the roads.

**Comment by kocio-pl** [comment]
> We currently do have the option where we violate 1.

So it means I'm a bit lost then. =}

> I don't think that would make a difference here. Rendering any buildings on top of highway areas would imply, assuming highway areas are rendered on top of roads, that those buildings are obscuring the roads.

I don't understand this case, could you explain it more and give an example?

**Comment by math1985** [comment]
We have three requirements:

1. building should be above highway-area-fill
If we violate this, the Eiffel tower square is rendered on top of the Eiffel tower. This is the current situation.

2. roads-casing should be above building
If we violate this, we have two options:
2a. Render both roads-casing and roads-fill below building
As roads are drawn wider than they are in reality (especially on low- and mid-zoom), the drawn roads often overlap with the buildings around the road. Rendering roads-fill below buildings would mean that buildings obscure the roads.
2b. Render both roads-casing below building, but roads-fill above it.
Same as above, except that the buildings now only obscure the casing of the road.

3. highway-area-fill should be above roads-casing
If we violate, we get ugly effects like in http://i.imgur.com/qTRO1yH.png.

And of course, 1, 2, and 3 are not compatible.

**Comment by dieterdreist** [comment]
2016-07-14 16:08 GMT+02:00 Matthijs Melissen <notifications@github.com>:

> We have three requirements:
>
>    1.
>
>    building should be above highway-area-fill
>    If we violate this, the Eiffel tower square is rendered on top of the
>    Eiffel tower. This is the current situation.
>
>

we should respect layers, because you can have highway areas above and
below buildings and it will not be solvable otherwise.

**Comment by gorn** [comment]
It is very well possible, that I am missing something very obvious, but I think there is **very easy solution**. I would call it 0th requirement, because if is the most important (IMHO).

0.  layer x+1 > layer x

     Or very briefly, respect layers if defined in data.

This would solve most of the problems. For the rest of the problems we can continue finding sophisticated solution (because the problem still exist if both object have the same layer). In particular it solves - Eifell Tower (and any other towers) and all undeground platforms (whenever mapped).

**Comment by math1985** [comment]
> layer x+1 > layer x

That unfortunately doesn't work. It would for example imply not render tunnels under fields, or not rendering railway tracks in stations covered by a roof.

Besides, there are also some technical issues that make this difficult to implement (there is no easy command to tell carto to first order by layer and then by everything else).

**Comment by dieterdreist** [comment]
2016-07-15 15:49 GMT+02:00 Matthijs Melissen <notifications@github.com>:

> That unfortunately doesn't work. It would for example imply not render
> tunnels under fields, or not rendering railway tracks in stations covered
> by a roof.



well, just the same as we render tunnels in a particular way (dashed), we
should care for other features that are below of something else. Respecting
the layer order does not necessarily mean that it's the same as the
rendering order, it means make a style decision to express the fact that
one thing is above the other. For rendering convenience it might be
required to add particular tags to these covered feature parts (e.g.
covered=yes).

**Comment by gorn** [comment]
> That unfortunately doesn't work. It would for example imply not render tunnels under fields, or not rendering railway tracks in stations covered by a roof.

Sorry I do not understand that. It used to be solved before when layer=-1 tunnel=yes was respected, so the same solution applies now.

**Comment by math1985** [comment]
> Sorry I do not understand that. 

You say that we should render layer 0 above layer -1. Fields are layer 0, tunnels are (or can be) layer -1. So fields are rendered on top of tunnels in your proposal, making the tunnel invisible.

**Comment by HolgerJeromin** [comment]
Note that we do not have transparent things, but adjusted (dashed)  tunnel roads rendered on top/afterwards.

**Comment by daganzdaanda** [comment]
I fear we are going in circles ;-)
The problems are known, possible solutions have been discussed. 
Now IMHO we need code and a few test renderings to see what's best. 
I'm really not capable of writing that code, and I'm afraid I don't have the time to start learning now... 
So - volunteers welcome!

**Comment by daganzdaanda** [comment]
@dieterdreist wrote...

> AFAIK you can't currently do vector based hatches in mapnik, but have to use the PolygonPatternSymbolizer to do it.

OK, so the PolygonPatternSymbolizer is what makes the hatches for prisons, military and allotments? If it can't do a cross-hatch, we could just do a fine linear hatching, maybe 2px wide and in some existing middle grey tone? 

It needs to be demo-ed to finetune, a lot of new visual problems could come from a fine hatching.

**Comment by gravitystorm** [comment]
> If you go for the filling, this is the docu, but maybe in mapnik 3 you can use svg (couldn't find this out): https://github.com/mapnik/mapnik/wiki/PolygonPatternSymbolizer

Yes, you can use SVGs for patterns in Mapnik3 (although there are various technical problems that make this tiresome to implement, trust me!)

**Comment by gorn** [comment]
@daganzdaanda  or anyone else - can you spot the changeset which caused the current situation?

**Comment by kocio-pl** [comment]
It was done here:

https://github.com/gravitystorm/openstreetmap-carto/pull/623

**Comment by eliotb** [comment]
Maybe related to #529 #2475 #2509 #2534 #2563 #2630 (my entry point to this)

It seems to me that the overall theme of all these issues is conflict between expectations for highway=something and area=yes when these two tags are combined.  Each of the issues may be complaining about a different specific tag (platform, pedestrian,...), or requesting a different solution (transparency vs render order)

A: I think that  the common expectation is that "for a given layer smaller areas will cover larger"small areas will render on top of larger areas" irrespective of area type (apart from layer != 0). ([1] paragraph 3).  

B: But on the other hand "roads are rendered over landcover" ([1] paragraph 1), and road-like areas are rendered after roads, (to cover up rounded ends of connecting ways?)

My preference is that A should take priority over B for areas.
It is really not obvious that if a building/monument/etc is placed in an area (of a particular type only) that a relation must be created, the tags of the area moved to the relation, the area added to relation as outer, the building added to relation as inner.  It is also a heap more work for the casual mapper.

[1] https://wiki.openstreetmap.org/wiki/Standard_tile_layer#Rendering_order

**Comment by eliotb** [comment]
A thought regarding this particular comment:

https://github.com/gravitystorm/openstreetmap-carto/issues/688#issuecomment-232675998
> 3. highway-area-fill should be above roads-casing
> If we violate, we get ugly effects like in http://i.imgur.com/qTRO1yH.png.

What if the full area is rendered 'lower' with other landuse etc, then a border of the area is rendered after the roads-casing?  This would negate the "ugly effects" mentioned above...

**Comment by eliotb** [comment]
Conceptually similar (IMO) to highway=pedestrian, area=yes are amenity=parking, and aeroway=apron.  Both of which do not obscure buildings etc that overlap them, even when all on layer=0

**Comment by sladen** [comment]
Yes, it is the 'casing' (outlines) that want rendering on top, *not* the flood fill.  This is the classic for tunnel rendering on maps.

**Comment by cwmke** [comment]
I'm having a similar issue with barriers not appearing in pedestrian/footway areas as shown below. It shows the difference in rendering between here and OSMAnd+ on Android.

![screenshot from 2018-03-19 13-13-58](https://user-images.githubusercontent.com/1923967/37614964-184c511a-2b7a-11e8-9b67-f94b8441acb1.png)
![screenshot_osmand _20180319-131103](https://user-images.githubusercontent.com/1923967/37614965-185e532e-2b7a-11e8-845d-5fb5eb3ccd95.png)

**Comment by kocio-pl** [comment]
This problem has its own ticket where you can discuss it: #528.

**Comment by cwmke** [comment]
Thanks @kocio-pl. Missed that one in the sea of similar issues :-).

**Comment by jragusa** [comment]
An other example:
https://www.openstreetmap.org/node/4492013017
![entrance_pedestrian](https://user-images.githubusercontent.com/9897203/46581919-3aeb0a80-ca40-11e8-8cfc-68eb5637e192.png)

**Comment by Lukas458** [comment]
Hi, I just wanted to ask whether I can help in this issue maybe?

**Comment by imagico** [comment]
This is a hard problem because there are several conflicting goals - as discussed above (specifically in https://github.com/gravitystorm/openstreetmap-carto/issues/688#issuecomment-201263916 and https://github.com/gravitystorm/openstreetmap-carto/issues/688#issuecomment-232675998 for example).

One possible approach to the problem was discussed in #3872.  This was neither rejected nor approved and could be re-visited.  But it comes with drawbacks.

**Comment by jeisenbe** [comment]
Re: https://github.com/gravitystorm/openstreetmap-carto/pull/3872 - that option might be more popular if we removed the highway=pedestrian casing or continued to render it below highway area fill.

**Comment by stragu** [comment]
Another example of area affected by this issue: https://www.openstreetmap.org/way/686546745#map=19/-27.49722/153.01578

The pedestrian area (`highway=pedestrian`) on the left covers a roof entirely (object selected), the area on the right (`highway=footway`) covers part of a roof.

Here is a screenshot of the issue:

![Screenshot of two areas showing the problem: roofs partially or entirely covered by `pedestrian` or `footway` areas.](https://user-images.githubusercontent.com/1747497/124060425-6031c980-da70-11eb-9a33-737f9ebc15b2.png)

Noting that the following renderings don't have that issue:
* CyclOSM
* Cycle Map
* Transport Map (although I feel it doesn't render the pedestrian/footway areas anyway)
* Humanitarian

**Comment by Sirorezka** [comment]
Hi,

Sorry, it's not clear. Why can't you downgrade "area=yes, highway=pedestrian" and "area=yes, highway=footpath" to level of "amenity=parking"? This will solve the issue.

**Comment by Sirorezka** [comment]
Also it's not clear why you regard the issue of "pedestrian areas" in bulk with "rendering all highway areas" and "rendering pedestrian ways".  This are three separate things, they should be splitted into three tasks.

**Comment by imagico** [comment]
> Sorry, it's not clear. Why can't you downgrade "area=yes, highway=pedestrian" and "area=yes, highway=footpath" (or same with multipolygon) to level of "amenity=parking"? This will solve the issue.

@matthijsmelissen already answered this in https://github.com/gravitystorm/openstreetmap-carto/issues/172#issuecomment-1603990058.  You can read up a more detailed discussion of this idea in #3872.

Also note by the way that pedestrian areas on top of buildings are probably similarly common in reality as pedestrian areas underneath buildings or `building=roof`.  So your 'solution' would (w.r.t. buildings, other issues are a different matter) just shift the problem.

> Also it's not clear why you regard the issue of "pedestrian areas" in bulk with "rendering all highway areas" and "rendering pedestrian ways".

`highway=pedestrian` polygons are considered together with other road polygons we render (in particular `highway=service` but also `highway=platform`/`railway=platform`) and linear `highway=pedestrian` because they are - both by us and by mappers - considered to be part of the road system and subject to the road mapping conventions, in particular with connectivity rules and the use of `tunnel=*`/`bridge=*`/`layer=*`.

**Comment by Sirorezka** [comment]
On your examples:
https://github.com/gravitystorm/openstreetmap-carto/issues/172#issuecomment-1603990058 - he mentioned problem with "road casing" (rounding last nodes for ways). I think this is much smaller issue compared to not visualising buildings, gardens and etc on top of pedestrian areas. Also I saw that in most related issues mappers ask to visualise tunnels and roads that go under the pedestrian area. 
https://github.com/gravitystorm/openstreetmap-carto/pull/3872 - is discussion about general rendering prioritizaiton. Not problem with pedestrian layer. Back to my point that you avoid looking at the pedestrian problem separately.

> Also note by the way that pedestrian areas on top of buildings are probably similarly common in reality as pedestrian areas underneath buildings or building=roof. So your 'solution' would (w.r.t. buildings, other issues are a different matter) just shift the problem.

I've mentioned this layering problem in my comment. See point about "7. No support of visualization prioritization for objects with different layer value". Completely agree with you on this, as of now you can't see gardens on top of buildings, swimming pools, parkings( !) and etc. This is also quite a huge issue. But if accepting that areas with assigned "layers" are not visualised accordingly then pedestrian on the roofs shouldn't be visible similarly to roof gardens, roof playing grounds, roof swimming pools, roof parkings and etc.

First of all this 'layer tag' visualization question should be a separate topic. The quick fix here might be also quite simple. First visualize everything with "layer=0" (or without 'layer' tag) accordingly to your prioritization schema. Then visualize "layer=1" without prioritization, after "layer=2" without prioritization and etc. If a mapper is using layers then he has enough experience to take into account difficulty of the area that he is plotting. So I can assure that he would assign layers to everything to prevent and fix any possible visualisation issues.

> highway=pedestrian polygons are considered together with other road polygons we render (in particular highway=service but also highway=platform/railway=platform) and linear highway=pedestrian because they are - both by us and by mappers - considered to be part of the road system and subject to the road mapping conventions, in particular with connectivity rules and the use of tunnel=*/bridge=*/layer=*.

Just to breakdown this topics: we are talking about wiki hierarchy and about carto hierarchy. I think they might be different. For example you are excluding some objects that are plotted by mappers (underground buildings and etc), but they are part of wiki hierachy. This is done because you have your own hierarchy of objects. As I said "highway=pedestrian" can be convereted to parking in carto hierarchy.

**Comment by imagico** [comment]
> https://github.com/gravitystorm/openstreetmap-carto/pull/3872 - is discussion about general rendering prioritizaiton.

No, it is not, this is the discussion of the idea of moving the rendering of road polygons from within the road layers to before them  (which is the same as what you proposed to 'solve' this issue).

As i wrote above:

> One possible approach to the problem was discussed in https://github.com/gravitystorm/openstreetmap-carto/pull/3872. This was neither rejected nor approved and could be re-visited. But it comes with drawbacks.

So if you want to pursue this idea again you need to look at the discussion we have had there and try to address the concerns raised.

To avoid misunderstanding: 'Road casing' refers to the darker outline we draw around the road fill at higher zoom levels.  For consistent results this needs to be drawn below the road fill but correctly ordered according to bridge/tunnel/layer tags to ensure proper display of where there is connectivity in the road system and where there is not.  More detailed discussion of this can be found [here](https://imagico.de/blog/en/navigating-the-maze-part-1/).

**Comment by jidanni** [comment]
All I know is in the case of #5050, if in a few hours one is going to City Hall of a population 3,000,000 city, to try to convince them about OpenStreetMap, whereas there will be no chance, there with the mayor and department heads, in the few minutes allowed, to tell them they need to click on alternative layers... to get the tunnel stuck on top of the roof of City Hall out of the way to avoid major embarrassment...

So in the few hours before the meeting, in our hotel room, after having tried various combinations of layer= and covered= etc. (per wiki advice) on both objects, one throws up their hands and following what we see above, unfortunately will have to zap the highway tag.

**Comment by mat-vc** [comment]
Issue has been reported and ongoing for 11 years. Is this on the timeline to be fixed? 
All other maps that I know of do render buildings in pedestrian areas fine. 
11 years...

**Comment by imagico** [comment]
11 years is not uncommon even for issues with much simpler solutions than this - see [here](https://imagico.de/blog/en/a-few-additional-comments-on-the-new-osm-carto-release/).

This issue is open, further work on the matter is welcome.  But an actual solution is hard and most likely technically awkward with our current framework - even more so if it should also address #2958 (see [here](https://imagico.de/blog/en/weaving-roads-2-creative-compositioning/) for a proof-of-concept of how this is possible).

What most other OSM based maps do is what has been suggested most recently in https://github.com/gravitystorm/openstreetmap-carto/issues/688#issuecomment-1603502150 and what we discussed in more depth in #3872.  But this is not an actual solution.  See there and above here regarding the problems of that approach.

**Comment by matthijsmelissen** [comment]
> Issue has been reported and ongoing for 11 years. Is this on the timeline to be fixed?
All other maps that I know of do render buildings in pedestrian areas fine.

This is a design decision by this style. See https://github.com/gravitystorm/openstreetmap-carto/issues/172, where the issue is explained in much detail.

In short, one might want three things:

1) buildings on top of road areas (like pedestrian squares)
2) roads on top of buildings
3) road areas on top of top of roads

Implementing all leads to a contradiction, so we can only have two out of three. This style has chosen not to implement 1. Many other styles have chosen not to implement 3 (they do not render road areas at all, or they don't render them similar to roads like we do, which allows them to render them below other roads).

I'm not saying this is a design decision that this style will never change, but we cannot solve it without abandoning requirement 2 or 3.

**Comment by imagico** [comment]
Indeed.  And for completeness: To get all three things with our current technical setup would require to have all roads and buildings in one layer.  That is what i called *technically awkward*.  I don't see this being viable here - but this is not a taboo - all options can be discussed.

**Comment by dieterdreist** [comment]
sent from a phone

> On 19 Jun 2025, at 22:43, Matthijs Melissen ***@***.***> wrote:
> 
> Implementing all leads to a contradiction, so we can only have two out of three.


the solution would be to respect layers, so you could handle all these cases if the data allows to determine the stacking. It is not easy to do though in this style, that’s true.

**Comment by matthijsmelissen** [comment]
> To get all three things with our current technical setup would require to have all roads and buildings in one layer.

No, I don’t think so. Even with a different setup, this would lead to a contradiction. This is a requirement problem, not an implementation problem.

> the solution would be to respect layers, so you could handle all these cases if the data allows to determine the stacking.

No, you don’t want to respect layers. For example, you don’t want tunnels under buildings to be fully hidden by the buildings on top of them.

**Comment by dieterdreist** [comment]
sent from a phone

> On 20 Jun 2025, at 00:03, Matthijs Melissen ***@***.***> wrote:
> 
> No, you don’t want to respect layers. For example, you don’t want tunnels under buildings to be fully hidden by the buildings on top of them.


of course not, by respecting layers I didn’t mean to always use OpenStreetMap layers as the rendering drawing order, rather show which layer is above which other layer, the technical drawing order then depends on this and the drawing style. I guess nobody is arguing that there is an actual problem with the current behavior and that the city hall rendering that was shown above is completely awkward.

With some help of the mappers (e.g. adding covered=yes and splitting partially covered areas where they are not covered) it could be possible to skip covered areas at first and later draw another layer for covered areas as dashed outlines above.

This style used to have similar issues also with tunnels in the past, in dense areas, when it drew the tunnels below stuff and they became hidden. Now this is changed and it’s the contrary: buildings are hid because of a tunnel below (another style decision: that fills are drawn for tunnels, and while they “simulate transparency” they actually are opaque and cover what is drawn below (but is  above in reality): https://www.openstreetmap.org/#map=17/41.940583/12.475292
because you don’t want semitransparent objects because they would make it unpredictable which colors result from blending)

Another similar issue are bridge areas, that currently are rendered below roads that run under them.

**Comment by imagico** [comment]
> No, I don’t think so. Even with a different setup, this would lead to a contradiction. This is a requirement problem, not an implementation problem.

My understanding is (and my thinking might be flawed on that of course) that the problem consists of two parts, one that is, indeed, a goal conflict as you describe, the other is actually solvable with a single consistent drawing order, but one that requires drawing some road elements before some buildings and some road elements after some buildings - hence the need to have all roads and buildings in one layer for that.

The latter type of problem is, for example, the *pedestrian area underneath building* (see https://github.com/gravitystorm/openstreetmap-carto/issues/5050#issue-2764948352 for an exemplary case) and the *pedestrian area on top of building* case.  This can be solved by consistent ordering based on secondary tags on the features in question.  That does not necessarily mean `layer=*`, it also could be things like `covered=yes`, `location=*`, `building:min_level=*` etc.

As a side effect solving this would also allow proper rendering of bridge/tunnel road polygons (which are currently rendered like surface road polygons) - see #2046.

The former type of problem, that is not solvable with one single globally consistent drawing order, is, for example, what we discussed in #528/#2958.  Here we have the problem that we'd want some linear features (waterways, barriers) to be rendered after the road polygons, so they show up when mapped over a pedestrian area for example, while we need to render them before linear roads to avoid messy results in well mapped areas.  At the same time we'd want to continue rendering the road polygons and linear roads together for properly visualizing connectivity with the separate rendering of casing and fill.

I have described this problem in detail [here](https://imagico.de/blog/en/weaving-roads-1-polygons-and-lines/) and presented a possible approach [here](https://imagico.de/blog/en/weaving-roads-2-creative-compositioning/) - which, however, goes beyond what is technically possible with our current framework.

**Comment by dieterdreist** [comment]
Ultimately, labels would also have to be dealt with, so that we don’t get labels of streetnames under a street rendered above, this can be particularly confusing with oneway arrows
